[1.8.9 -> 1.20.4] Partially converted, Exception in thread "main" java.lang.RuntimeException: java.nio.file.NoSuchFileException

``PS C:\Users\youar\Downloads\新しいフォルダー> java -jar ResourcePackConverter.jar --from 1.8.9 --to 1.20.4
Converting packs from: 1.8.9 to 1.20.4
Converting ResourcePack{path=.\fp7rism_8f163x8.zip}
  Running Converters
    Running NameConverter
    Failed to rename: cobblestone_mossy.png->mossy_cobblestone.png
    Failed to rename: double_plant_syringa_bottom.png->lilac_bottom.png
    Failed to rename: double_plant_syringa_top.png->lilac_top.png
    Failed to rename: grass_top.png->grass_block_top.png
    Failed to rename: piston_top_normal.png->piston_top.png
Failed to convert!
Exception in thread "main" java.lang.RuntimeException: java.nio.file.NoSuchFileException: .\fp7rism_8f163x8_converted\assets\minecraft\textures\entity\squid.png
        at com.agentdid127.resourcepack.library.utilities.Util.propagate(Util.java:102)
        at com.agentdid127.resourcepack.forwards.ForwardsPackConverter.runPack(ForwardsPackConverter.java:137)
        at com.agentdid127.resourcepack.forwards.ForwardsPackConverter.lambda$runDir$0(ForwardsPackConverter.java:145)
        at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
        at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:179)
        at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
        at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
        at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1845)
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
        at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
        at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:596)
        at com.agentdid127.resourcepack.forwards.ForwardsPackConverter.runDir(ForwardsPackConverter.java:145)
        at com.agentdid127.resourcepack.Main.main(Main.java:41)
Caused by: java.nio.file.NoSuchFileException: .\fp7rism_8f163x8_converted\assets\minecraft\textures\entity\squid.png
        at java.base/sun.nio.fs.WindowsException.translateToIOException(WindowsException.java:85)
        at java.base/sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:103)
        at java.base/sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:108)
        at java.base/sun.nio.fs.WindowsFileCopy.move(WindowsFileCopy.java:334)
        at java.base/sun.nio.fs.WindowsFileSystemProvider.move(WindowsFileSystemProvider.java:293)
        at java.base/java.nio.file.Files.move(Files.java:1432)
        at com.agentdid127.resourcepack.forwards.impl.NameConverter.convert(NameConverter.java:213)
        at com.agentdid127.resourcepack.forwards.ForwardsPackConverter.runPack(ForwardsPackConverter.java:129)
        ... 14 more``
        
 original 1.8.9 pack i tried to update to 1.20.4       
[fp7rism_8f163x8.zip](https://github.com/user-attachments/files/18205141/fp7rism_8f163x8.zip)

--- a/conversions/Forwards/src/main/java/com/agentdid127/resourcepack/forwards/impl/NameConverter.java
+++ b/conversions/Forwards/src/main/java/com/agentdid127/resourcepack/forwards/impl/NameConverter.java
@@ -16,10 +16,11 @@ import java.nio.file.Path;
 import java.nio.file.Paths;
 import java.util.Arrays;
 import java.util.List;
+import java.util.Objects;
 
 public class NameConverter extends Converter {
-    private int to;
-    private int from;
+    private final int to;
+    private final int from;
 
     private final Mapping blockMapping;
     private final Mapping newBlockMapping;
@@ -55,7 +56,7 @@ public class NameConverter extends Converter {
 
     /**
      * Fixes folder names and file names
-     * 
+     *
      * @param pack Resource Pack
      * @throws IOException Error handler
      */
@@ -76,12 +77,14 @@ public class NameConverter extends Converter {
                 Logger.debug("MCPatcher exists, switching to optifine");
                 if (minecraftPath.resolve("optifine").toFile().exists()) {
                     Logger.debug("OptiFine exists, merging directories");
-                    FileUtil.mergeDirectories(minecraftPath.resolve("optifine").toFile(),
-                            minecraftPath.resolve("mcpatcher").toFile());
-                } else
+                    FileUtil.mergeDirectories(minecraftPath.resolve("optifine").toFile(), minecraftPath.resolve("mcpatcher").toFile());
+                } else {
                     Files.move(minecraftPath.resolve("mcpatcher"), minecraftPath.resolve("optifine"));
-                if (minecraftPath.resolve("mcpatcher").toFile().exists())
+                }
+
+                if (minecraftPath.resolve("mcpatcher").toFile().exists()) {
                     FileUtil.deleteDirectoryAndContents(minecraftPath.resolve("mcpatcher"));
+                }
             }
 
             // 1.13 Models
@@ -89,16 +92,20 @@ public class NameConverter extends Converter {
             if (modelsPath.toFile().exists()) {
                 // 1.13 block/item name change
                 if (modelsPath.resolve("blocks").toFile().exists()) {
-                    if (modelsPath.resolve("block").toFile().exists())
+                    if (modelsPath.resolve("block").toFile().exists()) {
                         FileUtil.deleteDirectoryAndContents(modelsPath.resolve("block"));
+                    }
+
                     Files.move(modelsPath.resolve("blocks"), modelsPath.resolve("block"));
                 }
 
                 // Update all blocks for 1.13
                 renameAll(blockMapping, ".json", modelsPath.resolve("block"));
                 if (modelsPath.resolve("items").toFile().exists()) {
-                    if (modelsPath.resolve("item").toFile().exists())
+                    if (modelsPath.resolve("item").toFile().exists()) {
                         FileUtil.deleteDirectoryAndContents(modelsPath.resolve("item"));
+                    }
+
                     Files.move(modelsPath.resolve("items"), modelsPath.resolve("item"));
                 }
 
@@ -110,25 +117,29 @@ public class NameConverter extends Converter {
                     Path itemModelPath = modelsPath.resolve("iem");
                     renameAll(newItemMapping, ".json", itemModelPath);
                     if (!itemModelPath.resolve("ink_sac.json").toFile().exists()
-                            && itemModelPath.resolve("black_dye.json").toFile().exists())
-                        Files.copy(itemModelPath.resolve("black_dye.json"),
-                                itemModelPath.resolve("ink_sac.json"));
+                            && itemModelPath.resolve("black_dye.json").toFile().exists()) {
+                        Files.copy(itemModelPath.resolve("black_dye.json"), itemModelPath.resolve("ink_sac.json"));
+                    }
+
                     if (!itemModelPath.resolve("cocoa_beans.json").toFile().exists()
-                            && itemModelPath.resolve("brown_dye.json").toFile().exists())
-                        Files.copy(itemModelPath.resolve("brown_dye.json"),
-                                itemModelPath.resolve("cocoa_beans.json"));
+                            && itemModelPath.resolve("brown_dye.json").toFile().exists()) {
+                        Files.copy(itemModelPath.resolve("brown_dye.json"), itemModelPath.resolve("cocoa_beans.json"));
+                    }
+
                     if (!itemModelPath.resolve("bone_meal.json").toFile().exists()
-                            && itemModelPath.resolve("white_dye.json").toFile().exists())
-                        Files.copy(itemModelPath.resolve("white_dye.json"),
-                                itemModelPath.resolve("bone_meal.json"));
+                            && itemModelPath.resolve("white_dye.json").toFile().exists()) {
+                        Files.copy(itemModelPath.resolve("white_dye.json"), itemModelPath.resolve("bone_meal.json"));
+                    }
+
                     if (!itemModelPath.resolve("lapis_lazuli.json").toFile().exists()
-                            && itemModelPath.resolve("blue_dye.json").toFile().exists())
-                        Files.copy(itemModelPath.resolve("blue_dye.json"),
-                                itemModelPath.resolve("lapis_lazuli.json"));
+                            && itemModelPath.resolve("blue_dye.json").toFile().exists()) {
+                        Files.copy(itemModelPath.resolve("blue_dye.json"), itemModelPath.resolve("lapis_lazuli.json"));
+                    }
                 }
 
-                if (to > Util.getVersionProtocol(packConverter.getGson(), "1.19"))
+                if (to > Util.getVersionProtocol(packConverter.getGson(), "1.19")) {
                     renameAll(blockMapping19, ".json", modelsPath.resolve("block"));
+                }
 
                 if (to > Util.getVersionProtocol(packConverter.getGson(), "1.20.3")) {
                     renameAll(itemMapping203, ".json", modelsPath.resolve("item"));
@@ -137,10 +148,10 @@ public class NameConverter extends Converter {
             }
 
             // Update BlockStates
-            Path blockStates = pack.getWorkingPath()
-                    .resolve("assets/minecraft/blockstates".replace("/", File.separator));
-            if (blockStates.toFile().exists())
+            Path blockStates = pack.getWorkingPath().resolve("assets/minecraft/blockstates".replace("/", File.separator));
+            if (blockStates.toFile().exists()) {
                 renameAll(blockMapping, ".json", blockStates);
+            }
 
             // Update textures
             Path texturesPath = pack.getWorkingPath().resolve("assets/minecraft/textures".replace("/", File.separator));
@@ -170,21 +181,24 @@ public class NameConverter extends Converter {
                     renameAll(newItemMapping, ".png.mcmeta", itemPath);
                     Path itemModelPath = modelsPath.resolve("item");
                     if (!itemModelPath.resolve("ink_sac.png").toFile().exists()
-                            && itemModelPath.resolve("black_dye.png").toFile().exists())
-                        Files.copy(itemModelPath.resolve("black_dye.png"),
-                                itemModelPath.resolve("ink_sac.png"));
+                            && itemModelPath.resolve("black_dye.png").toFile().exists()) {
+                        Files.copy(itemModelPath.resolve("black_dye.png"), itemModelPath.resolve("ink_sac.png"));
+                    }
+
                     if (!itemModelPath.resolve("cocoa_beans.png").toFile().exists()
-                            && itemModelPath.resolve("brown_dye.png").toFile().exists())
-                        Files.copy(itemModelPath.resolve("brown_dye.png"),
-                                itemModelPath.resolve("cocoa_beans.png"));
+                            && itemModelPath.resolve("brown_dye.png").toFile().exists()) {
+                        Files.copy(itemModelPath.resolve("brown_dye.png"), itemModelPath.resolve("cocoa_beans.png"));
+                    }
+
                     if (!itemModelPath.resolve("bone_meal.png").toFile().exists()
-                            && itemModelPath.resolve("white_dye.png").toFile().exists())
-                        Files.copy(itemModelPath.resolve("white_dye.png"),
-                                itemModelPath.resolve("bone_meal.png"));
+                            && itemModelPath.resolve("white_dye.png").toFile().exists()) {
+                        Files.copy(itemModelPath.resolve("white_dye.png"), itemModelPath.resolve("bone_meal.png"));
+                    }
+
                     if (!itemModelPath.resolve("lapis_lazuli.png").toFile().exists()
-                            && itemModelPath.resolve("blue_dye.png").toFile().exists())
-                        Files.copy(itemModelPath.resolve("blue_dye.png"),
-                                itemModelPath.resolve("lapis_lazuli.png"));
+                            && itemModelPath.resolve("blue_dye.png").toFile().exists()) {
+                        Files.copy(itemModelPath.resolve("blue_dye.png"), itemModelPath.resolve("lapis_lazuli.png"));
+                    }
                 }
 
                 // Entities
@@ -193,11 +207,18 @@ public class NameConverter extends Converter {
                 // 1.16 Iron golems
                 if (from < Util.getVersionProtocol(packConverter.getGson(), "1.15")
                         && to >= Util.getVersionProtocol(packConverter.getGson(), "1.15")) {
-                    if (!entityPath.resolve("iron_golem").toFile().exists())
+                    if (!entityPath.resolve("iron_golem").toFile().exists()) {
                         entityPath.resolve("iron_golem").toFile().mkdir();
-                    if (entityPath.resolve("iron_golem.png").toFile().exists())
-                        Files.move(entityPath.resolve("iron_golem.png"), entityPath
-                                .resolve("iron_golem/iron_golem.png"));
+                    }
+
+                    if (entityPath.resolve("iron_golem.png").toFile().exists()) {
+                        Path newIronGolemPath = entityPath.resolve("iron_golem/iron_golem.png");
+                        if (newIronGolemPath.toFile().exists()) {
+                            newIronGolemPath.toFile().delete();
+                        }
+
+                        Files.move(entityPath.resolve("iron_golem.png"), newIronGolemPath);
+                    }
                 }
 
                 // 1.17 Squid
@@ -207,21 +228,38 @@ public class NameConverter extends Converter {
                     renameAll(itemMapping17, ".png", texturesPath.resolve("item"));
                     renameAll(blockMapping17, ".png", modelsPath.resolve("block"));
                     renameAll(itemMapping17, ".png", modelsPath.resolve("item"));
-                    if (!entityPath.resolve("squid").toFile().exists())
-                        entityPath.resolve("squid").toFile().mkdir();
-                    if (!entityPath.resolve("squid.png").toFile().exists())
-                        Files.move(entityPath.resolve("squid.png"),
-                                entityPath.resolve("squid/squid.png"));
+
+                    Path squidFolderPath = entityPath.resolve("squid");
+                    if (!squidFolderPath.toFile().exists()) {
+                        squidFolderPath.toFile().mkdir();
+                    }
+
+                    if (entityPath.resolve("squid.png").toFile().exists()) {
+                        Path newSquidPath = squidFolderPath.resolve("squid.png");
+                        if (newSquidPath.toFile().exists()) {
+                            newSquidPath.toFile().delete();
+                        }
+
+                        Files.move(entityPath.resolve("squid.png"), newSquidPath);
+                    }
                 }
 
                 if (to >= Util.getVersionProtocol(packConverter.getGson(), "1.19")
-                        && from < Util.getVersionProtocol(packConverter.getGson(), "1.19"))
+                        && from < Util.getVersionProtocol(packConverter.getGson(), "1.19")) {
                     renameAll(blockMapping19, ".png", blockPath);
+                }
 
                 // 1.13 End Crystals
                 if (entityPath.resolve("endercrystal").toFile().exists()
-                        && !entityPath.resolve("end_crystal").toFile().exists())
-                    Files.move(entityPath.resolve("endercrystal"), entityPath.resolve("end_crystal"));
+                        && !entityPath.resolve("end_crystal").toFile().exists()) {
+                    Path newEnderCrystalPath = entityPath.resolve("end_crystal");
+                    if (newEnderCrystalPath.toFile().exists()) {
+                        newEnderCrystalPath.toFile().delete();
+                    }
+
+                    Files.move(entityPath.resolve("endercrystal"), newEnderCrystalPath);
+                }
+
                 findEntityFiles(entityPath);
             }
         }
@@ -229,33 +267,33 @@ public class NameConverter extends Converter {
 
     /**
      * Finds files in entity folder
-     * 
+     *
      * @param path
      * @throws IOException
      */
     protected void findEntityFiles(Path path) throws IOException {
-        if (!path.toFile().exists())
-            return;
-        File directory = path.toFile();
-        for (File file : directory.listFiles()) {
-            if (!file.isDirectory())
-                continue;
-            renameAll(entityMapping, ".png", file.toPath());
-            renameAll(entityMapping, ".png.mcmeta", file.toPath());
-            findEntityFiles(file.toPath());
+        if (path.toFile().exists()) {
+            File directory = path.toFile();
+            for (File file : Objects.requireNonNull(directory.listFiles())) {
+                if (file.isDirectory()) {
+                    renameAll(entityMapping, ".png", file.toPath());
+                    renameAll(entityMapping, ".png.mcmeta", file.toPath());
+                    findEntityFiles(file.toPath());
+                }
+            }
         }
     }
 
     /**
      * Finds files in folders called
-     * 
+     *
      * @param path
      * @throws IOException
      */
-    protected void findFiles(Path path) throws IOException {
+    protected void findFiles(Path path) {
         if (path.toFile().exists()) {
             File directory = path.toFile();
-            for (File file : directory.listFiles()) {
+            for (File file : Objects.requireNonNull(directory.listFiles())) {
                 if (file.isDirectory()) {
                     if (file.getName().equals("items")) {
                         Logger.debug("Found Items folder, renaming");
@@ -270,11 +308,13 @@ public class NameConverter extends Converter {
                     findFiles(file.toPath());
                 }
 
-                if (file.getName().contains("("))
+                if (file.getName().contains("(")) {
                     FileUtil.renameFile(path.resolve(file.getName()), file.getName().replaceAll("[()]", ""));
+                }
 
-                if (!file.getName().equals(file.getName().toLowerCase()))
+                if (!file.getName().equals(file.getName().toLowerCase())) {
                     Logger.debug("Renamed: " + file.getName() + "->" + file.getName().toLowerCase());
+                }
 
                 FileUtil.renameFile(path.resolve(file.getName()), file.getName().toLowerCase());
             }
@@ -283,7 +323,7 @@ public class NameConverter extends Converter {
 
     /**
      * Renames folder
-     * 
+     *
      * @param mapping
      * @param extension
      * @param path
@@ -296,7 +336,7 @@ public class NameConverter extends Converter {
             List<String> grasses = Arrays.asList("grass", "tall_grass", "double_grass");
             if (from <= Util.getVersionProtocol(packConverter.getGson(), "1.12.2")) {
                 if ((path.endsWith("blockstates") || path.endsWith("textures/block"))) {
-                    grasses.stream().forEach(name -> {
+                    grasses.forEach(name -> {
                         String newName = mapping.remap(name);
                         Boolean ret = FileUtil.renameFile(Paths.get(path + File.separator + name + extension),
                                 newName + extension);
@@ -314,35 +354,44 @@ public class NameConverter extends Converter {
             // remap snow jsons, but not images.
             if (from < Util.getVersionProtocol(packConverter.getGson(), "1.13")
                     && to >= Util.getVersionProtocol(packConverter.getGson(), "1.13")) {
-                if (path.resolve("snow_layer.json").toFile().exists())
+                if (path.resolve("snow_layer.json").toFile().exists()) {
                     FileUtil.renameFile(path.resolve("snow_layer" + extension), "snow" + extension);
+                }
             }
 
+            Logger.addTab();
             Files.list(path).forEach(path1 -> {
                 if (!path1.toString().endsWith(extension))
                     return;
 
-                String baseName = path1.getFileName().toString().substring(0,
-                        path1.getFileName().toString().length() - extension.length());
+                String baseName = path1.getFileName().toString().substring(0, path1.getFileName().toString().length() - extension.length());
 
                 // skip the already renamed grass blocks
-                if (grasses.contains(baseName)
-                        && (path.endsWith("blockstates") || path.endsWith("textures/block")))
+                if (grasses.contains(baseName) && (path.endsWith("blockstates") || path.endsWith("textures/block"))) {
                     return;
+                }
 
-                String newName = mapping.remap(baseName);
-                if (newName != null && !newName.equals(baseName)) {
-                    Boolean ret = FileUtil.renameFile(path1, newName + extension);
-                    if (ret == null)
+                String remappedName = mapping.remap(baseName);
+                if (remappedName != null && !remappedName.equals(baseName)) {
+                    Path newPath = path1.getParent().resolve(remappedName + extension);
+                    if (newPath.toFile().exists()) {
+                        // Same note from Slicer.java line 38
+                        newPath.toFile().delete();
+                    }
+
+                    Boolean renameSuccess = FileUtil.renameFile(path1, newPath);
+                    if (renameSuccess == null) {
                         return;
-                    if (ret) {
-                        Logger.debug("Renamed: " + path1.getFileName().toString() + "->" + newName + extension);
-                    } else if (!ret) {
-                        Logger.log("Failed to rename: " + path1.getFileName().toString() + "->" + newName
-                                + extension);
+                    }
+
+                    if (renameSuccess) {
+                        Logger.debug("Renamed: " + path1.getFileName().toString() + "->" + newPath.getFileName());
+                    } else if (!renameSuccess) {
+                        Logger.log("Failed to rename: " + path1.getFileName().toString() + "->" + newPath.getFileName());
                     }
                 }
             });
+            Logger.subTab();
         }
     }
 

--- a/conversions/Forwards/src/main/java/com/agentdid127/resourcepack/forwards/impl/textures/slicing/Slicer.java
+++ b/conversions/Forwards/src/main/java/com/agentdid127/resourcepack/forwards/impl/textures/slicing/Slicer.java
@@ -1,9 +1,5 @@
 package com.agentdid127.resourcepack.forwards.impl.textures.slicing;
 
-import java.io.IOException;
-import java.lang.reflect.Method;
-import java.nio.file.Path;
-
 import com.agentdid127.resourcepack.forwards.impl.textures.SlicerConverter;
 import com.agentdid127.resourcepack.library.utilities.FileUtil;
 import com.agentdid127.resourcepack.library.utilities.ImageConverter;
@@ -12,12 +8,16 @@ import com.agentdid127.resourcepack.library.utilities.Logger;
 import com.google.gson.Gson;
 import com.google.gson.JsonObject;
 
+import java.io.IOException;
+import java.lang.reflect.Method;
+import java.nio.file.Path;
+
 public class Slicer {
-    public static <T> void runSlicer(Gson gson, Slice slice, Path root, Class<T> predicateClass, int from,
-            boolean texture_has_metadata)
+    public static <T> void runSlicer(Gson gson, Slice slice, Path root, Class<T> predicateClass, int from, boolean has_metadata)
             throws IOException {
-        if (!SlicerConverter.PredicateRunnable.run(gson, from, slice.getPredicate()))
+        if (!SlicerConverter.PredicateRunnable.run(gson, from, slice.getPredicate())) {
             return;
+        }
 
         Path path = root.resolve(slice.getPath());
         if (!path.toFile().exists()) {
@@ -28,11 +28,19 @@ public class Slicer {
         ImageConverter converter = new ImageConverter(slice.getWidth(), slice.getHeight(), path);
 
         Path imagePath = path;
-        if (slice.getPathName() != null)
+        if (slice.getPathName() != null) {
             imagePath = path.resolveSibling(slice.getPathName());
+        }
 
         for (Texture texture : slice.getTextures()) {
             Path texturePath = root.resolve(texture.getPath());
+            if (texturePath.toFile().exists()) {
+                // Sometimes packs include both newer and older textures but
+                // since we are upgrading to a newer version, we want what was shown in the original
+                // version so we delete the existing new one, so we can replace it with the
+                // version used in said version.
+                texturePath.toFile().delete();
+            }
             FileUtil.ensureParentExists(texturePath);
 
             if (predicateClass != null) {
@@ -48,7 +56,6 @@ public class Slicer {
 
             try {
                 Box box = texture.getBox();
-
                 converter.saveSlice(
                         box.getX(),
                         box.getY(),
@@ -56,19 +63,20 @@ public class Slicer {
                         box.getHeight(),
                         texturePath);
 
-                if (texture.shouldRemove())
+                if (texture.shouldRemove()) {
                     converter.fillEmpty(
                             box.getX(),
                             box.getY(),
                             box.getWidth(),
                             box.getHeight());
+                }
 
-                if (texture_has_metadata) {
+                if (has_metadata) {
                     JsonObject metadata = texture.getMetadata();
-                    if (metadata.keySet().isEmpty() || metadata.entrySet().isEmpty())
-                        continue;
-                    Path metadataPath = texturePath.resolveSibling(texturePath.getFileName() + ".mcmeta");
-                    JsonUtil.writeJson(gson, metadataPath, metadata);
+                    if (!metadata.keySet().isEmpty() && !metadata.entrySet().isEmpty()) {
+                        Path metadataPath = texturePath.resolveSibling(texturePath.getFileName() + ".mcmeta");
+                        JsonUtil.writeJson(gson, metadataPath, metadata);
+                    }
                 }
             } catch (Exception exception) {
                 Logger.log("Failed to slice texture '" + texture.getPath() + "' (error='"
@@ -78,10 +86,12 @@ public class Slicer {
             }
         }
 
-        if (!path.toFile().delete())
+        if (!path.toFile().delete()) {
             Logger.debug("Failed to remove '" + path.getFileName() + "' whilst slicing.");
+        }
 
-        if (!slice.shouldDelete())
+        if (!slice.shouldDelete()) {
             converter.store(imagePath);
+        }
     }
 }

--- a/library/src/main/java/com/agentdid127/resourcepack/library/utilities/FileUtil.java
+++ b/library/src/main/java/com/agentdid127/resourcepack/library/utilities/FileUtil.java
@@ -5,31 +5,26 @@ import java.io.IOException;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.util.Comparator;
+import java.util.Objects;
 
 public final class FileUtil {
     public static void moveIfExists(Path in, Path out) throws IOException {
-        if (!in.toFile().exists())
-            return;
-        ensureParentExists(out);
-        Files.move(in, out);
-    }
-
-    public static void copyIfExists(Path in, Path out) throws IOException {
-        if (!in.toFile().exists())
-            return;
-        ensureParentExists(out);
-        Files.copy(in, out);
+        if (in.toFile().exists()) {
+            ensureParentExists(out);
+            Files.move(in, out);
+        }
     }
 
     public static void ensureParentExists(Path path) {
         Path parentPath = path.getParent();
-        if (!parentPath.toFile().exists())
+        if (!parentPath.toFile().exists()) {
             parentPath.toFile().mkdirs();
+        }
     }
 
     /**
      * Copies Directory
-     * 
+     *
      * @param src  directory Source
      * @param dest Directory Destination
      * @throws IOException
@@ -48,23 +43,22 @@ public final class FileUtil {
 
     /**
      * Deletes full directory
-     * 
+     *
      * @param dirPath Path of Directory to delete
      * @throws IOException
      */
     public static void deleteDirectoryAndContents(Path dirPath) throws IOException {
-        if (!dirPath.toFile().exists())
-            return;
-
-        // noinspection ResultOfMethodCallIgnored
-        Files.walk(dirPath).sorted(Comparator.reverseOrder()).map(Path::toFile).forEach(file -> {
-            file.delete();
-        });
+        if (dirPath.toFile().exists()) {
+            // noinspection ResultOfMethodCallIgnored
+            Files.walk(dirPath).sorted(Comparator.reverseOrder()).map(Path::toFile).forEach(file -> {
+                file.delete();
+            });
+        }
     }
 
     /**
      * If file exists, return file with correct casing.
-     * 
+     *
      * @param path
      * @return
      * @throws IOException
@@ -77,9 +71,26 @@ public final class FileUtil {
      * @return null if file doesn't exist, {@code true} if successfully renamed,
      *         {@code false} if failed
      */
-    public static Boolean renameFile(Path file, String newName) {
-        if (!file.toFile().exists())
+    public static Boolean renameFile(Path file, Path newFile) {
+        if (!file.toFile().exists()) {
             return null;
+        }
+        try {
+            Files.move(file, newFile);
+            return true;
+        } catch (IOException e) {
+            return false;
+        }
+    }
+
+    /**
+     * @return null if file doesn't exist, {@code true} if successfully renamed,
+     *         {@code false} if failed
+     */
+    public static Boolean renameFile(Path file, String newName) {
+        if (!file.toFile().exists()) {
+            return null;
+        }
         try {
             Files.move(file, file.getParent().resolve(newName));
             return true;
@@ -90,16 +101,18 @@ public final class FileUtil {
 
     /**
      * Takes dir2 and merges it into dir1
-     * 
+     *
      * @param dir1
      * @param dir2
      */
-    public static Boolean mergeDirectories(File dir1, File dir2) throws IOException {
-        if (!dir1.exists() && !dir2.exists())
-            return null;
+    public static void mergeDirectories(File dir1, File dir2) throws IOException {
+        if (!dir1.exists() && !dir2.exists()) {
+            return;
+        }
 
         String targetDirPath = dir1.getAbsolutePath();
         File[] files = dir2.listFiles();
+        assert files != null;
         for (File file : files) {
             if (file.isDirectory()) {
                 Logger.debug(targetDirPath + File.separator + file.getName());
@@ -113,9 +126,8 @@ public final class FileUtil {
             }
         }
 
-        if (dir2.list().length == 0)
+        if (Objects.requireNonNull(dir2.list()).length == 0) {
             Files.delete(dir2.toPath());
-
-        return true;
+        }
     }
 }

